reqh/reqh_service.h

```diff
@@ -237,6 +237,8 @@ struct m0_reqh_service {
 	unsigned                           rs_level;
 	/** Key for per-locality-per-svc fom count. */
 	int                                rs_fom_key;
+	/** Fom's posted to queueit() AST */
+	struct m0_atomic64                 rs_fom_queued;
```

Comment:
Instead of atomic, please use a per-locality locker, similar to ->rs_fom_key.

Reply:
I have tried this option, as locality locker update (get and set) is not atomic, latest incremented value in m0_fom_queue() is not available during queueit().

``3043080.911.426144  7ffe62061080  DEBUG   fop/fom.c:1350:m0_fom_locality_queueit_ast_inc      fom:0x1a21f68, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73
3043080.911.432854  7f02c7ffeb40  DEBUG   fop/fom.c:1363:m0_fom_locality_queueit_ast_dec      fom:0x1a1ae78, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73


3043080.911.437234  7ffe62061080  DEBUG   fop/fom.c:1350:m0_fom_locality_queueit_ast_inc      fom:0x1a29058, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73
3043080.911.447732  7ffe62061080  DEBUG   fop/fom.c:1350:m0_fom_locality_queueit_ast_inc      fom:0x1a30148, service:0x198d6b0, **cnt:9** loc:0x15e55b0, key:73
3043080.911.448304  7f02c7ffeb40  DEBUG   fop/fom.c:1363:m0_fom_locality_queueit_ast_dec      fom:0x1a21f68, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73

Here in queueit() before the value of the counter is older one, so missing to decrement some foms, which causes issue in finalisation.

3043080.911.457722  7ffe62061080  DEBUG   fop/fom.c:1350:m0_fom_locality_queueit_ast_inc      fom:0x1a37238, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73
3043080.911.463580  7f02c7ffeb40  DEBUG   fop/fom.c:1363:m0_fom_locality_queueit_ast_dec      fom:0x1a30148, service:0x198d6b0, **cnt:8** loc:0x15e55b0, key:73

```

Currently locker update is not atomic, so can't use this method.